### ISSUE TITLE
Mocks cannot handle floating point?

### ISSUE THREAD

**Issue by awilke** [cgreen-devs/cgreen#96]
I'm having trouble getting mocks to work for functions which use floating point values.
1. When a mocked function is expected to return a floating point value, the mock actually returns the integer version of that value.
2. When a parameter to a mocked function has an expected value, that mock will segfault.

When running the attached code example [fp_mock_test.txt](https://github.com/cgreen-devs/cgreen/files/551321/fp_mock_test.txt):

```
Running "floating point mock test" (2 tests)...
fp_mock_test.c:15: Failure: foobar_returns_floating_point 
        Expected [test_func()] to [equal double] [282.4] within [8] significant figures
                actual value:   282.000000
                expected value: 282.400000

fp_mock_test.c:24: Exception: foobar_param_doesnt_segfault 
        Test terminated with signal: Segmentation fault: 11

Completed "floating point mock test": 1 failure, 1 exception in 113ms.
```

Am I using mocks incorrectly here, or do mocks not currently support floating point types? This is using the latest Cgreen @ bead2ca, on OS X 10.11.6, with `gcc version 4.9.3 (Homebrew gcc49 4.9.3)`.

**Comment by thoni56** [comment]
Thank you for an excellent issue report! It was easy to expose the problem you have.

I sure shouldn't segv. I'll see if I get some time in the next few days to have a look at it.

**Comment by thoni56** [comment]
Seems like `mock()` can't handle doubles, like you expected.

In the first test (`foobar_returns_floating_point`) the `will_return()` always handles it's argument as an `int`. We probably need a `will_return_double()` and a `create_return_double_value_constraint()`, unless someone has a clever way to store a double in an `intptr_t`... (and I think that's why we have the `BoxedDouble`...)

The segmentation fault in the second test (`foobar_param_doesnt_segfault`) is caused by `mock()` trying to evaluate a constraint for a double when the actual value (the argument to `mock()`) has not been stored as one.

You can use `box_double()` to force boxing the double argument to match what the constraint expects:

    uint8_t foobar_param(const double bob) {
        return (uint8_t)mock(box_double(bob));
    }

To make this more user-friendly we should make `mock()` box the double before passing it on as an argument to the evaluation of the constraints. And since there might be more arguments than one, we can't have a `mock_double()` to signal that the argument(s) should be boxed doubles. (We might need `mock_double()` to unbox a double return value, though. Haven't explored that yet...)

### Note to developers:
Maybe we need to add a mechanism to indicate the type of the value in the constraints. When everything could be represented as a very big `int` we could always compare to that and values were always stored the same way. With support for doubles, the value stored in the constraint might also be  boxed double, which is a pointer. Reading an integer to be a pointer does mostly not work as expected, as this issue shows.

If a user forgets a `_double` somewhere, or the `box_double()` in a `mock()`-call, we could well end up in this situation even though we have provided additional macros/functions that fixes the issues presented here. A "type indicator" in the constraints would allow us to catch such user mistakes. Or even an abstract `Value` type union with a tag:

    typedef struct {
        ValueType type;
        union {
            double double_value;
            ...
        }
    }

**Comment by thoni56** [comment]
I've created a [gist for this](https://gist.github.com/thoni56/6ba7b78b14aee0fc4015b4ab40f40e73#file-96-c).

**Comment by thoni56** [comment]
And I've updated the [gist](https://gist.github.com/thoni56/6ba7b78b14aee0fc4015b4ab40f40e73#file-96-c) with a hacked version which passes all the tests. It shows where boxing and unboxing needs to happen to make `mock()` and `will_return()` handle doubles correctly.

This is maybe a crude stop-gap for some, but it's very likely not forward compatible, depending on how we decide to solve this.

**Comment by awilke** [comment]
Thanks for the update and workaround!

It seems that the boxing/unboxing workaround does not work if a mocked function takes one double argument, and any other arguments (doesn't have to be double). [This updated gist](https://gist.github.com/awilke/da55ed0e32059d37f8c1c49728e1e3d1) results in more segfaults:

```
Running "floating point mock test" (4 tests)...
./96_multi_param.c:43: Failure: foobar_multi_param_doesnt_segfault 
        Expected [[alice] parameter in [foobar_multi_param]] to [equal] [27U]
                actual value:                   [0]
                expected value:                 [27]

./96_multi_param.c:37: Exception: foobar_multi_param_doesnt_segfault 
        Test terminated with signal: Segmentation fault: 11

./96_multi_param.c:54: Exception: foobar_multi_fp_param_doesnt_segfault 
        Test terminated with signal: Segmentation fault: 11

Completed "floating point mock test": 5 passes, 1 failure, 2 exceptions in 9ms.
```

But the workaround did work for me with functions which only had one, double argument.

**Comment by matthargett** [comment]
We tried numerous ways of doing this a few years ago, and the best we came up with was overloading the mock_ function to take an automatic template parameter, kind of like mockitopp. You could still test flat C code , but the test code would be compiled as C++.

Happy to remote pair to flesh out the idea.

**Comment by thoni56** [comment]
I'm in the process of exploring a humungous refactoring, replacing `intptr_t` and `BoxedDouble` for `expected_value` (and further on also `actual_value`) with

    typedef enum {INTEGER, STRING, DOUBLE, POINTER} CgreenValueType;
    
    typedef struct {
        CgreenValueType type;
        union {
            intptr_t integer_value;
            double double_value;
            void *pointer_value;
            const char *string_value;
        } value;
    } CgreenValue;

In a way overloading C value types. So far I've refactored all `create_constraint...` and `compare...` and so far it seems like a feasible route. This should work since passing and returning structs had been in the C standard since C89.

Happy to take any input on this strategy.

**Comment by JDeeth** [comment]
Hi all, very new to Cgreen here. Apologies, I don't completely understand the above discussion, but I think this is related perhaps, using plain float values without a function?

```
Ensure(Foo, float_comparison) {
  assert_that(3.2, is_equal_to_double(3.2));
}
```

This generates a compiler error, for instance with Clang:
`/usr/local/include/cgreen/internal/cpp_assertions.h:26: error: reinterpret_cast from 'double' to 'const std::string *' (aka 'const basic_string<char> *') is not allowed`

I get the same error using `is_equal_to` as well.

Interim workaround is to write `assert_that( foo == 3.2 );` instead.

**Comment by thoni56** [comment]
@JDeeth, sorry, we should probably take the internal design discussions off line to avoid confusion.

Your problem is kind of related to the "double" issue, but there is a simple fix for your issue. You should use `assert_that_double()` instead of `assert_that()`.

One goal in this issue to be able to give better error messages for this type of mistakes, which are easy to make with the current design.

**Comment by thoni56** [comment]
I have a PR that will be merged soon which fixes the `segv` problems discovered/reported by @awilke with multiple `box_double()`. At least this will make the intended way to handle this work.

But, there are a few more things to do
- document the handling of double in a clear and visible way
- handling `box_double()` automatically, if possible
- handling mis-matched or missed 'box/unbox_double()` in a more graceful way than 'segmentation violation'...

**Comment by awilke** [comment]
Sweet, thanks. I tried out that fix and my previous issues are resolved when manually using `box_double()` and `unbox_double()`. The floating-point mocks work as expected.

Another issue to note is that `always_expect()` does not work with those same floating-point mocks. At runtime, I get something like
```
malloc: *** error for object 0x7fcc22600000: pointer being freed was not allocated
```
which is to be expected when the mock's `unbox_double()` is being called multiple times for a single `box_double()` in the `always_expect()`.

**Comment by thoni56** [comment]
Thanks for trying the fix out, and for the heads up on `always_expect()`. Will have a look at that too.

**Comment by thoni56** [comment]
Should be fixed by #118
